Hosting Dashboard: Add domain glue records add/edit screen

m1r0 opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of DOTDASH-202

## Proposed Changes

* This change introduces the glue records add/edit form.
* A few things are missing that will be added in the next iteration. Those are tracking, tests, and probably better error handling.

<img width="715" height="369" alt="image" src="https://github.com/user-attachments/assets/bfc07ddb-6dcf-410e-9dc8-03d7995e2beb" />

<img width="714" height="365" alt="image" src="https://github.com/user-attachments/assets/56cf5229-028b-4ac9-a9fd-57f3ebef1ca5" />

https://github.com/user-attachments/assets/f8eaec71-6e28-4bc1-9987-9e2971f57685


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* This is part of the [Hosting Dashboard: Domains](https://linear.app/a8c/project/hosting-dashboard-domains-7a553b8d4a37/overview) project.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to `/v2/domains/[domain]/glue-records` and click `Add glue record`.
* You should be redirected to the add screen.
* Test the form validation by adding an invalid server name and IP address.
* Add a valid glue record. You should be redirected back to the listing screen.
* Click on the new record. You should be redirected to the edit screen.
* Update the IP address. You should be redirected to the listing screen again.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

m1r0 committed:
Add glue records creation form

m1r0 committed:
Fix text case

m1r0 committed:
Add glue records edit form

m1r0 committed:
Switch to `ValidatedInputControl` for the name server input

This allows for adding the needed suffix.

m1r0 committed:
Add a variable for the suffix value

m1r0 committed:
Fix the placeholder value to make it consistent

m1r0 commented on client/dashboard/domains/overview-glue-records/form.tsx:
```diff
@@ -0,0 +1,143 @@
+// eslint-disable-next-line no-restricted-imports
+import { ValidatedInputControl } from '@automattic/components/src/validated-form-controls/components/input-control';
```
It looks like the `ValidatedInputControl` is still in beta, so I'm not sure if this is the right thing to do.

I'm using this because I found no other component that has a validated input that can have a suffix.

@hambai, I saw that you faced the same issue in https://github.com/Automattic/wp-calypso/pull/105113. Do you think this is an ok approach for now?

p-jackson replied:
Is that what this [message was referring to](https://github.com/Automattic/wp-calypso/pull/105113#:~:text=fix%20it%20upstream-,Show%20the%20domain%20as%20suffix%20in%20the%20subdomain%20field%20%2D%20seems%20like%20this%20can%20be%20achieved%20by%20using%20the%20InputControl%20component%20instead%20of%20TextControl%20but%20I%27m%20not%20sure%20how%20to%20do%20that.%20It%20seems%20that%20InputControl%20is%20supposed%20to%20replace%20the%20TextControl%20so%20I%27ll%20ask%20around%20to%20figure%20out%20what%20our%20options%20are,-Hide%20the%20%22Source)?

If it is something we need for domain input boxes, maybe this is something we could create and contribute upstream 🤔 A place to ask about it might be #atelier-data-views

We could leave the suffix out of the component for now and follow up once we have a done the component work?

arthur791004 replied:
I don't think we can import it directly...so I didn't use it in https://github.com/Automattic/wp-calypso/pull/105191

m1r0 replied:
Thanks, Arthur. I've [updated](https://github.com/Automattic/wp-calypso/pull/105213/commits/2f593989b6f6d9c28e9cd7f5c5e05f559fec9a1c) the code to use `InputControl` while we wait for a better solution.

@mirka, I'm not sure if someone has already contacted you, but do you know if there are plans to release the `validated-form-controls` package? Our project could use the `ValidatedInputControl` component.

mirka replied:
Thanks for the ping! Those validation components are kind of meant to be a temporary solution, especially now that we're working on a new set of components. Hence the reluctance to make them directly available to consumers.

I also just recently removed them from the wp-calypso repo entirely (#105219), since canonical development has already moved to the Gutenberg repo.

At the moment we're [exposing](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataform--validation) them for external use mainly through the `DataForm` abstraction in `@wordpress/dataviews`. Is that an option here?

m1r0 replied:
Thanks for the quick response, @mirka!

> At the moment we're [exposing](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataform--validation) them for external use mainly through the DataForm abstraction in @wordpress/dataviews. Is that an option here?

Our main hurdle is not having the ability to add a suffix to the `DataForm` fields and also have validation. In our current project, we have a few forms that look something like this:

<img width="647" height="242" alt="image" src="https://github.com/user-attachments/assets/ecd39d03-1dfc-4f15-958b-4a06fa8d4cbf" />

One option that we found was to use the `InputControl` component, which has a suffix, but we can't make the validation messages work. `ValidatedInputControl` does the job, but we can't use that because it's not exposed. 

Do you have any suggestions on how to approach this? 🙏 Maybe contribute a solution, e.g., adding a field suffix in Gutenberg?

mirka replied:
I see. Not sure how far DataForm wants to support extension points like this in the standard data types (cc @oandregal).

If not feasible to add this feature in DataForm, I'm fine with temporarily exposing any of the validated form components you need as an unlocked API in `@automattic/components`, similar to [what we did](https://github.com/Automattic/wp-calypso/blob/a2eb55a63e20cc40730a6d60df1e15e5d3b0a476/packages/components/src/menu/index.tsx#L29) with `Menu`. (In `Menu` we went through a lot of trouble to manually add back the TS information that [gets lost](https://github.com/WordPress/gutenberg/pull/46131#issuecomment-1415812737) when components are locked through the `privateApis` mechanism, but I think we can skip that effort for this. At the time, we were expecting to be in a contribution ban for a much longer time.)

oandregal replied:
> I see. Not sure how far DataForm wants to support extension points like this in the standard data types.

I can see how it'd be beneficial to support this, so consumers can configure the Edit component.

My main concern is that we shouldn't couple DataForm to a specific component implementation. We want to be future-proof, which means that we can update the underlying components when/if require it. I don't think we should allow passing _any_ prop to the underlying components, for example. This needs to be discussed in a case-by-case basis.

@mirka do you think prefix/suffix for inputs is a common enough concept that we can incorporate it to DataForm and be future-proof?

I'm thinking we could do this via the Edit function of a field. It's now a React component or a string (e.g., `text`, `email`, `integer`). It could also be an object with some props to be passed down to the component. For example:

```js
{
  id: 'field',
  Edit: {
    type: 'text',
    suffix: 'suffix-will-go-here'
  }
}
```

mirka replied:
I totally understand how you wouldn't want to couple DataForm to underlying component implementations, so personally I'm not pushing too hard in that direction.

And the validated form components will be available for public use eventually, in one form or another, so in the long run it's not absolutely necessary that DataForm support these customizations in the standard field types.

On the future-proof angle, the prefix/suffix props themselves are pretty stable, and there are no plans to remove them even in the next-gen components. However, there's a subcomponent (`InputControlPrefixWrapper`) that can be used to control padding in the prefix/suffix slots, and that part is subject to change 😅 Plus those old subcomponents will not work with the next-gen `InputControl`s. So these types of small back compat issues cannot be 100% avoided if we go down this route. If this is the first time this request has come up, it's maybe too early to say that this is a common enough use case.

m1r0 replied:
> I'm fine with temporarily exposing any of the validated form components you need as an unlocked API in @automattic/components

@mirka, just to make sure I understand, our option right now is to unlock `ValidatedInputControl` and wait for the validated form components to become public. Is that correct?

mirka replied:
Correct 👍

m1r0 replied:
Got it, thank you!

m1r0 commented on client/dashboard/domains/overview-glue-records/utils.ts:
```diff
@@ -0,0 +1,22 @@
+	}
+
+	return true;
+}
```
The validation logic is copied from the [previous implementation](https://github.com/Automattic/wp-calypso/blob/dfe8f71280eeca0c777391fb22efb39b7552d43e/client/my-sites/domains/domain-management/settings/cards/glue-records-card.tsx#L163-L177).

p-jackson commented on client/dashboard/app/queries/domain-glue-records.ts:
```diff
@@ -8,6 +14,22 @@ export const domainGlueRecordsQuery = ( domainName: string ) =>
+		mutationFn: ( glueRecord: DomainGlueRecord ) => updateDomainGlueRecord( glueRecord ),
+		onSuccess: () => {
+			queryClient.invalidateQueries( domainGlueRecordsQuery( domainName ) );
+		},
```
The response body from the server will be passed to `onSuccess`. Is it possible to use this response body to update the data in the query client immediately before invalidating the `domainGlueRecordsQuery()`?

In the video you can see that it takes a second before the updated value pops in, and that is because the UI doesn't update until the query invalidation is complete. Ideally the UI could update immediately since we know the server has responded with a success.

Same thing for the record creation success handler.

You update the client using `queryClient.setQueryData()`. Here's an example:
https://github.com/Automattic/wp-calypso/blob/9268e6290f8c13e55f682753481525215f7dd1b4/client/dashboard/app/queries/site-jetpack-module.ts#L25

m1r0 replied:
Oh, that's good to know. Thanks for the explanation!

I've updated all mutations in https://github.com/Automattic/wp-calypso/pull/105213/commits/2db8665b239c3ce52130efb3c7783c6aa9ea21bc.

p-jackson commented on client/dashboard/data/domain-glue-records.ts:
```diff
@@ -12,6 +12,33 @@ export function fetchDomainGlueRecords( domainName: string ): Promise< DomainGlu
+	);
+}
+
+export function updateDomainGlueRecord( glueRecord: DomainGlueRecord ): Promise< void > {
```
I see that the update function is returning `void`, so perhaps it is not possible to use the data back from the server to update `queryClient`. But you can also use the data passed to the mutation function to update the `queryClient` too.

m1r0 replied:
Good pointer! 🙇

p-jackson commented on client/dashboard/domains/overview-glue-records/add.tsx:
```diff
@@ -0,0 +1,51 @@
+import type { FormData } from './form';
+
+export default function AddDomainGlueRecords() {
+	const router = useRouter();
```
You don't need the whole router here, maybe `useNavigate()`?

m1r0 replied:
Makes sense. 👍

Updated in https://github.com/Automattic/wp-calypso/pull/105213/commits/ca1c252ac754b8f56ca79b7a580efd637ee4a008.

p-jackson commented on client/dashboard/domains/overview-glue-records/edit.tsx:
```diff
@@ -0,0 +1,73 @@
+				createSuccessNotice( __( 'Glue record updated successfully.' ), {
+					type: 'snackbar',
+				} );
+				router.navigate( {
```
Same here

p-jackson commented on client/dashboard/domains/overview-glue-records/edit.tsx:
```diff
@@ -0,0 +1,73 @@
+export default function EditDomainGlueRecords() {
+	const router = useRouter();
+	const { domainName, nameServer } = domainRoute.useParams();
+	const { data: glueRecordsData } = useQuery( domainGlueRecordsQuery( domainName ) );
```
Since we're already using loading this query in the router we can use React Suspense here. That way we don't have to deal with the case where `glueRecordsData` is undefined.

```suggestion
	const { data: glueRecordsData } = useSuspenseQuery( domainGlueRecordsQuery( domainName ) );
```

m1r0 replied:
Yep, makes sense. Updated in https://github.com/Automattic/wp-calypso/pull/105213/commits/8002ca0846c82ad75e0b6838e045b245e629c439.

p-jackson commented on client/dashboard/domains/overview-glue-records/edit.tsx:
```diff
@@ -0,0 +1,73 @@
+
+	const existingGlueRecord = useMemo( () => {
+		if ( ! glueRecordsData || ! nameServer ) {
+			throw notFound();
```
Using `useMemo` for navigation is a little suspicious. To cause a navigation you want to throw directly within the render function body. I see `domain-forwarding` is using a memo here, maybe that's where this came from 🤔 

In this case it looks like it is possible to know whether you should redirect before you even get to the component, so if you want you could do the redirecting at the router level? Like this sort of thing https://github.com/Automattic/wp-calypso/blob/9268e6290f8c13e55f682753481525215f7dd1b4/client/dashboard/app/router.tsx#L147

m1r0 replied:
Good point. I've moved the redirect logic to the router in https://github.com/Automattic/wp-calypso/pull/105213/commits/61c95f88603d6cd062013e7cea8b39fce491c565.

> I see domain-forwarding is using a memo here, maybe that's where this came from

Nailed it. 😅

p-jackson commented on client/dashboard/domains/overview-glue-records/utils.ts:
```diff
@@ -0,0 +1,22 @@
+export function isValidNameServer( nameServer: string ): boolean {
```
Do we need a separate utils file for this? Or could it live in the existing `utils/domains` file if it is things that are generally true about domains outside the scope of the component.

m1r0 replied:
I agree, the utils could be useful outside of the scope of the component. 👍 Move them in https://github.com/Automattic/wp-calypso/pull/105213/commits/bd938c2b97ea3e23d9923b419fbbeb70bbc55381.

arthur791004 commented on client/dashboard/domains/overview-glue-records/add.tsx:
```diff
@@ -0,0 +1,51 @@
+		const glueRecord = {
+			nameserver: formData.nameServer,
+			ip_addresses: [ formData.ipAddress ],
+		};
```
Would it make more sense to handle this inside the `DomainGlueRecordsForm` component? Otherwise, we’ll need to duplicate the transform in two places.

m1r0 replied:
Makes sense, thanks! I've cleaned it up in https://github.com/Automattic/wp-calypso/pull/105213/commits/80e42f0a80c69b99d9aff301675b9f6c3c313918.

arthur791004 commented on client/dashboard/domains/overview-glue-records/form.tsx:
```diff
@@ -0,0 +1,143 @@
+								variant="primary"
+								type="submit"
+								isBusy={ isSubmitting }
+								disabled={ isSubmitting }
```
I think we should disable the button when the input is invalid

p-jackson replied:
This feels like something that would be in a design spec. E.g. another way it could be handled is keeping the button enabled so that you can show an error.

But I see the dataform validation storybook disables the button when the form is invalid. So maybe that counts as an official design system stance 😅 https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataform--validation

m1r0 replied:
If that's the guideline, here's the update https://github.com/Automattic/wp-calypso/pull/105213/commits/079da51015e5bff161401f2d7c26ed722aec89c2. 🙂 

Appreciate your help!

m1r0 committed:
Update the client before invalidating the queries

m1r0 committed:
Replace `useRoute` with `useNavigate`

m1r0 committed:
Switch to the `useSuspenseQuery` hook

fushar reviewed (commented):
Could we rename `domains/overview-glue-records` to `domains/domain-glue-records`, to be consistent with the others?

p-jackson commented on client/dashboard/app/queries/domain-glue-records.ts:
```diff
@@ -8,10 +14,47 @@ export const domainGlueRecordsQuery = ( domainName: string ) =>
+export const domainGlueRecordCreateMutation = ( domainName: string ) =>
+	mutationOptions( {
+		mutationFn: ( glueRecord: DomainGlueRecord ) => createDomainGlueRecord( glueRecord ),
+		onSuccess: ( newData: unknown, createdGlueRecord: DomainGlueRecord ) => {
```
Great updating the `queryClient` on success 👍 
Another heads up that we should not provide data types for the `onSuccess` function. The `mutationOptions()` helper makes sure these types are correct. The `newData` will automatically be whatever type `mutationFn` returns, and `createdGlueRecord` will automatically be whatever type `mutationFn` accepts.

I've found it is best to let TanStack Query infer the types where it can because it's actually easy to mistype them. When I first rolled out the `mutationOptions` and `queryOptions` helpers I found a couple of runtime bugs because of incorrect types 😁 


```suggestion
		onSuccess: ( newData, createdGlueRecord ) => {
```

Same advice for the other `onSuccess` handlers.

You do have to supply a type for the `mutationFn` parameter though. There's no way for TanStack to know what you want this type to be (but don't worry because TS will complain about it if you forget)

m1r0 replied:
Thanks for teaching me the ropes, Philip! 🙇

The types were [updated](https://github.com/Automattic/wp-calypso/pull/105213/commits/76dee31974bb26c6797ae0e44dbd802a65a2edab).

p-jackson reviewed (commented):
I left some more feedback but after that I think the code is looking good from my point of view.

I think when it comes to that input field which includes the subdomain it would be better to just leave it like Arthur did in the earlier PR. https://app.graphite.dev/github/pr/Automattic/wp-calypso/105213/Hosting-Dashboard-Add-domain-glue-records-add%2Fedit-screen#comment-PRRC_kwDOArzw_86ICrFp

It seems like we should figure it out upstream and then we can bulk-fix-up the Hosting Dashboard input fields that need it.

m1r0 committed:
Move the util helpers

m1r0 committed:
Fix missing `useMemo` dependency

m1r0 committed:
Cleanup glue record transform duplication

m1r0 committed:
Rename `overview-glue-records` to `domain-glue-records`

m1r0 committed:
Disable the submit button when form is invalid

m1r0 committed:
Replace `ValidatedInputControl` with `InputControl`

The `ValidatedInputControl` component is still in beta, so we decided not to use it.
This means that the input won't show error messages for now.

m1r0 committed:
Fix domain router imports

m1r0 committed:
Improve mutation types

m1r0 committed:
Move the 404 redirect to the router

m1r0 committed:
Fix missing useMemo dependencies

leonardost reviewed (approved):
Tested and LGTM! I think the code looks great. I have some nitpicks but none are blockers, I think we can deploy this and iterate.

<img width="704" height="352" alt="Screenshot 2025-08-22 at 11 18 13" src="https://github.com/user-attachments/assets/bfcd9eb6-535b-4be4-abed-a9b55d0d9bb0" />

Maybe we could have a "Cancel" button to go back to the glue records list page?

<img width="706" height="362" alt="Screenshot 2025-08-22 at 11 19 08" src="https://github.com/user-attachments/assets/c8dc7ba9-125c-4cbe-8132-b05dd65adcd8" />

This list shouldn't be too long (domains who use glue records usually have 2 or 3), so I wonder if we need to show the filtering and display options here.

<img width="699" height="343" alt="Screenshot 2025-08-22 at 13 42 00" src="https://github.com/user-attachments/assets/2aaaeaa3-5951-4996-b02c-f279c414704b" />

The suffix part in this input looks weird for long domain names, but I guess we'll change these components soon.

m1r0 commented:
> I think we can deploy this and iterate.

Sounds good. Thanks for the review, Leo!

oandregal commented on client/dashboard/domains/domain-glue-records/form.tsx:
```diff
@@ -0,0 +1,160 @@
+						/>
+					);
+				},
+				isValid: {
+					required: true,
+					custom: ( item ) => {
+						if ( ! isValidNameServerSubdomain( item.nameServer ) ) {
+							return __( 'Please enter a valid name server.' );
+						}
+						return null;
+					},
+				},
```
Hey, I just want to make sure we are aware that the rules provided here are not doing anything because they're not handled in the custom `Edit` control.

Edit: see how it's implemented in the [input control for DataForm](https://github.com/WordPress/gutenberg/blob/c436c1daa417a54cb4033fc42106b1a9b7bf1df8/packages/dataviews/src/dataform-controls/text.tsx#L40) and the [custom Edit example](https://github.com/WordPress/gutenberg/blob/c436c1daa417a54cb4033fc42106b1a9b7bf1df8/packages/dataviews/src/components/dataform/stories/index.story.tsx#L437) in the storybook.

m1r0 replied:
Thanks, André. I'll refactor this part in a future PR.
